Added a utility package for helping with fetching database connections.

sbose78 opened the pull request:
Why:
There are many parts in the code where fetching a database connection is needed, especially in integration tests. It gets messy if we have the code to do the same replicated across multiple methods.

What:
Added a new utils/ package to organize the code. Also added a test which currently is nothing more than a sample usage.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/almighty/almighty-core/112)
<!-- Reviewable:end -->

sbose78 commented:
@aslakknutsen 
Could you please review my approach? Tried hard to make the methods look clean :)

bartoszmajsak commented on utils/database_connection_helper.go:
```diff
@@ -0,0 +1,44 @@
+
+	db, err := gorm.Open(databaseConfig.databaseName, fmt.Sprintf(databaseConfig.connectionString))
+
+	defer db.Close()
```
Why `defer` here? Isn't it too early for that?

kwk replied:
@bartoszmajsak you're right: too early.

sbose78 replied:
Oops didnt notice I forgot to remove that. Doing it right away! @bartoszmajsak

sbose78 replied:
This is what confused me from main.go

```
db, err := gorm.Open("postgres", fmt.Sprintf("host=%s user=postgres password=mysecretpassword sslmode=disable", dbHost))
	if err != nil {
		panic("failed to connect database: " + err.Error())
	}
	defer db.Close()

	// Migrate the schema
	migration.Perform(db)
```

sbose78 replied:
This design of ```defer``` is along the lines of the way async calls are done in C#.

kwk commented on utils/database_connection_helper.go:
```diff
@@ -0,0 +1,44 @@
+
+// DetectDatabaseHost figures out the db host. The code needs to be modified based on our deployment strategy.
+func detectDatabaseHost() string {
+	dbHost := os.Getenv("DBHOST")
```
I would not solely rely on `DBHOST` environment variables but use something like [viper](https://github.com/spf13/viper).

tsmaeder commented:
Is the purpose of this package to open a database? If so, why do we need to expose so any things as API? Why not just have a single "Connect" method that return a db connection?
if you call the package "database", client code would look like

db, err:= db.Connect()

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[utils/database_connection_helper.go, line 29 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/112#-KOu0uM1MRCfwdjfDlqx:-KOu0uM2ItxgvwdmIbi1:b-6fp9ig) ([raw file](https://github.com/almighty/almighty-core/blob/4a915eb5ac435207f07033ce52d39da8bbbac0a3/utils/database_connection_helper.go#L29)):*
> ```Go
> func DetectConnectionString() string {
> 	host := detectDatabaseHost()
> 	connectionString := fmt.Sprintf("host=%s user=postgres password=mysecretpassword sslmode=disable", host)
> ```

I'm pretty sure this doesn't compile...where is the return statement?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/112#-:-KOu161N09MmxpiaoIUn:bzhymsy)*
<!-- Sent from Reviewable.io -->

sbose78 commented:
- updated the name to db/connection.go and db/connection_test.go
- moved  out the ```defer connection.Close()``` statement.
- fixed the missed return statement.

aslakknutsen commented on db/connection.go:
```diff
@@ -0,0 +1,43 @@
+	"os"
+
+	"github.com/jinzhu/gorm"
+	_ "github.com/lib/pq"
```
This can't be here if 'postgres' is an optional thing. Main or Test control that.

sbose78 replied:
Right, this would end up running the init method which should not be done if we are using a non-postgres database.

aslakknutsen commented:
Hmm.. How did the 'util' v of something that was 3 lines of code end up as 49 lines of support code and is now 6 lines to use?

aslakknutsen commented:
How about we add some thing like database.DefaultConnect() ?

sbose78 commented:
> Is the purpose of this package to open a database? If so, why do we need to expose so any things as API? Why not just have a single "Connect" method that return a db connection?
> if you call the package "database", client code would look like
> 
> db, err:= db.Connect()

Primarily to allow the connection string to be modified and/or with a different database. Basically allowing it to be overridden ( not in the OOP way ).

sbose78 commented:
Thanks @aslakknutsen

> Hmm.. How did the 'util' v of something that was 3 lines of code end up as 49 lines of support code and is now 6 lines to use?

Not really, because it isn't just 3 lines of code when the same piece of code is replicated in integration tests. 

```
connectionString := DetectConnectionString()
databaseName := DetectDatabaseName()
dc := DatabaseConfiguration{databaseName, connectionString}
connection, err := dc.GetDatabaseConnection()
```

The above is clean from 
- connection strings, 
- database name strings , 
- logic to check os env variable.
- The problem: Too many lines :| 

This could have been smaller had I been able to assign default values to method parameters in go :(   ( I'm missing that language feature from Python )

I wish to keep the *getting the default connection* simple yet give options to override.

> How about we add some thing like database.DefaultConnect() ?
Sounds good, how should I add options to explicitly specify the database creds when not using the default? ( Or do we need that at all? )

sbose78 commented:
As per review comments:
- removed the configurability of database name since we would be using postgres in the recent future.

Only connection string is configurable now.
